src/dask_awkward/lib/optimize.py

```diff
@@ -178,70 +102,80 @@ def optimize_columns(dsk: HighLevelGraph, keys: Sequence[Key]) -> HighLevelGraph
-        touch_data(arg)
-
-
-def _mock_output(layer):
-    """Update a layer to run the _touch_all_data."""
-    assert len(layer.dsk) == 1
-
-    new_layer = copy.deepcopy(layer)
-    mp = new_layer.dsk.copy()
-    for k in iter(mp.keys()):
-        mp[k] = (_touch_all_data,) + mp[k][1:]
-    new_layer.dsk = mp
-    return new_layer
+    dsk2 = dsk.layers.copy()
+
+    lays = {_[0] for _ in keys if isinstance(_, tuple)}
+    all_reps = set()
+    for ln in lays:
+        if ln in dsk.layers and hasattr(dsk.layers[ln], "meta"):
+            m = dsk.layers[ln].meta
+            if not isinstance(m, ak._nplikes.typetracer.MaybeNone):
+                # maybenone cases should already have been all touched
+                # but we could extract the .content here
+                touch_data(m)
+            rep = getattr(dsk.layers[ln].meta, "_report", ())
+            if rep:
+                all_reps.update(rep)
+    name = tokenize("output", lays)
+    [_.commit(name) for _ in all_reps]
+    all_layers = tuple(dsk.layers) + (name,)
+
+    for k, lay, cols in _optimize_columns(dsk.layers, all_layers):
+        new_lay = lay.project(cols)
+        dsk2[k] = new_lay
+
+    return HighLevelGraph(dsk2, dsk.dependencies)
+
+
+def _optimize_columns(dsk, all_layers):
+    for k, lay in dsk.copy().items():
+        if not isinstance(lay, AwkwardInputLayer) or not hasattr(
+            lay.io_func, "_column_report"
+        ):
+            continue
+        rep = lay.io_func._column_report
+        cols = set()
+        # this loop not required after next ak release
+        for ln in all_layers:
+            try:
+                cols.update(rep.data_touched_in((ln,)))
+            except KeyError:
+                pass
+            try:
+                cols.update(rep.shape_touched_in((ln,)))
+            except KeyError:
+                pass
+        yield k, lay, cols
+
+
+def necessary_columns(*args):
+    dsk = {}
+    all_reps = set()
+    all_layers = set()
+    for arg in args:
+        dsk.update(arg.dask.layers)
+        all_layers.update(arg.dask.layers)
+        if hasattr(arg, "_meta"):  # isinstance(, (dak.Scalar, dak.Array)?
+            touch_data(arg._meta)
+            rep = getattr(arg.dask.layers[arg.name].meta, "_report", ())
+            if rep:
+                all_reps.update(rep)
+    name = tokenize("output", args)
+    [_.commit(name) for _ in all_reps]
+    all_layers = tuple(all_layers) + (name,)
+
+    out = {}
+    for k, _, cols in _optimize_columns(dsk, all_layers):
+        first = (_buf_to_col(s) for s in cols)
+        # remove root "offsets", which appears when computing divisions
+        second = {s for s in first if s and s != "offsets"}
+        # remove columns that have sub-columns also in the set
+        for c in second.copy():
+            if "." in c:
+                parent = c.rsplit(".", 1)[0]
+                second.discard(parent)
+        out[k] = sorted(second)
```

Comment:
This logic assumes that buffer names are `.` delimited and are hierarchical, i.e. `foo.bar` is below `foo`. This isn't true for form remapping, which can move arbitrary buffers around. 

That is to say, the real operation we want to implement is `necessary_buffers` which happens to be `necessary_columns` for the Parquet/JSON schemes.